### ISSUE TITLE
3.8.1 : collect + sort query issued through ArangoJS crashes server (sigsegv)

### ISSUE THREAD

**Issue by matcho** [arangodb/arangodb#14949]
## My Environment

* __ArangoDB Version__:        3.8.1
* __Storage Engine__:             RocksDB
* __Deployment Mode__:       Single Server
* __Deployment Strategy__:   systemd
* __Infrastructure__:               own
* __Operating System__:        Ubuntu 20.04
* __Total RAM in your machine__:        64GB (machine 1) / 256 GB (machine 2)
* __Disks in use__:        HDD (machine 1) / SSD (machine 2)
* __Used Package__:             Ubuntu .deb

## Component, Query & Data

__Affected feature__:
ArangoDB server stability / AQL query with ArangoJS driver

__AQL query (if applicable)__:
```
let p = @p
collect project = p
sort project
return 1
```
with `p` a bind variable defined in Javascript code.

__Dataset__:
Not relevant: query does not read any collection data.

## Steps to reproduce

[3.8.1_crash_test.zip](https://github.com/arangodb/arangodb/files/7389020/3.8.1_crash_test.zip)

1. unpack test program above
2. `npm install`
3. run the Nodejs/ArangoJS test program many times, for ex. using `while true; do nodejs test.js ; done`
4. after a random number of executions, one will eventually fail and make the server crash

ArangoJS version is 7.6.0 (latest) but bug already occurred with 7.5.0 
ArangoDB version is 3.8.1

This bug is reproducible on another machine running ArangoDB v3.8.1 on Ubuntu 20.04 too

__Problem__:
Server crashes with an error like the following (strack trace is not always the same):
```
2021-10-21T10:18:31Z [132603] FATAL [a7902] {crash} 💥 ArangoDB 3.8.1 [linux], thread 4 [SchedWorker] caught unexpected signal 11 (SIGSEGV) accessing address 0x0000000000000000: signal handler invoked - image base address: 0x00007f03ee1bf000 - CPU context: rip: 0x00007f03ef50655f, rsp: 0x00007f03e61f75c0, efl: 0x0000000000010246, rbp: 0x00007f03e61f75d0, rsi: 0x00007f03e61f75e8, rdi: 0x00007f03e61f75f0, rax: 0x00007f03ef506510, rbx: 0x0000000000000000, rcx: 0x00007f02c10ae600, rdx: 0x0000000000000000, r8: 0x0000000000000003, r9: 0x0000000000000000, r10: 0x00000000000000b8, r11: 0x00007f03e61fab20, r12: 0x00007f03e61f75f0, r13: 0x00007f03e61f75e8, r14: 0x0000000000000000, r15: 0x00007f03e61f7be8
2021-10-21T10:18:31Z [132603] INFO [c962b] {crash} Backtrace of thread 4 [SchedWorker]
2021-10-21T10:18:31Z [132603] INFO [308c3] {crash} frame  1 [+0x0000000000806515] (anonymous namespace)::crashHandlerSignalHandler(int, siginfo_t*, void*) (+0x35)
2021-10-21T10:18:31Z [132603] INFO [308c3] {crash} frame  2 [+0x00000000039bf393] sigprocmask (+0x21)
2021-10-21T10:18:31Z [132603] INFO [308c3] {crash} frame  3 [+0x000000000134f81a] arangodb::aql::ExecutionPlan::findCollectionAccessVariables() (+0x6a)
2021-10-21T10:18:31Z [132603] INFO [308c3] {crash} frame  4 [+0x00000000013b49e5] arangodb::aql::ExecutionEngine::instantiateFromPlan(arangodb::aql::Query&, arangodb::aql::ExecutionPlan&, bool, arangodb::aql::SerializationFormat) (+0x4d5)
2021-10-21T10:18:31Z [132603] INFO [308c3] {crash} frame  5 [+0x00000000006ab9b7] arangodb::aql::Query::prepareQuery(arangodb::aql::SerializationFormat) [clone .constprop.0] (+0x7d7)
2021-10-21T10:18:31Z [132603] INFO [308c3] {crash} frame  6 [+0x00000000015c2235] arangodb::aql::Query::execute(arangodb::aql::QueryResult&) (+0xa95)
2021-10-21T10:18:31Z [132603] INFO [308c3] {crash} frame  7 [+0x00000000006d29cc] arangodb::RestCursorHandler::processQuery(bool) [clone .constprop.0] (+0x2c)
2021-10-21T10:18:31Z [132603] INFO [308c3] {crash} frame  8 [+0x0000000001952afe] arangodb::RestCursorHandler::registerQueryOrCursor(arangodb::velocypack::Slice const&) (+0x6fe)
2021-10-21T10:18:31Z [132603] INFO [308c3] {crash} frame  9 [+0x000000000195381b] arangodb::RestCursorHandler::execute() (+0x8fb)
2021-10-21T10:18:31Z [132603] INFO [308c3] {crash} frame 10 [+0x00000000019c01a6] arangodb::rest::RestHandler::executeEngine(bool) (+0xb6)
2021-10-21T10:18:31Z [132603] INFO [308c3] {crash} frame 11 [+0x00000000019ea72a] arangodb::rest::RestHandler::runHandlerStateMachine() (+0x36a)
2021-10-21T10:18:31Z [132603] INFO [308c3] {crash} frame 12 [+0x0000000001670b16] arangodb::Scheduler::WorkItem<arangodb::rest::CommTask::handleRequestSync(std::shared_ptr<arangodb::rest::RestHandler>)::{lambda()#1}>::invoke() (+0x126)
2021-10-21T10:18:31Z [132603] INFO [308c3] {crash} frame 13 [+0x0000000001859e2f] arangodb::SupervisedScheduler::runWorker() (+0x43f)
2021-10-21T10:18:31Z [132603] INFO [308c3] {crash} frame 14 [+0x000000000080d1a9] arangodb::Thread::startThread(void*) (+0x69)
2021-10-21T10:18:31Z [132603] INFO [308c3] {crash} frame 15 [+0x00000000007b135f] ThreadStarter(void*) [clone .lto_priv.0] (+0x8f)
2021-10-21T10:18:31Z [132603] INFO [308c3] {crash} frame 16 [+0x00000000039c6167] start (+0x70)
2021-10-21T10:18:31Z [132603] INFO [ded81] {crash} available physical memory: 67364077568, rss usage: 5692440576, vsz usage: 17729486848, threads: 47

```

Client receives error:
```
Error: socket hang up
    at connResetException (node:internal/errors:683:14)
    at Socket.socketOnEnd (node:_http_client:471:23)
    at Socket.emit (node:events:406:35)
    at endReadableNT (node:internal/streams/readable:1329:12)
    at processTicksAndRejections (node:internal/process/task_queues:83:21) {
  code: 'ECONNRESET',
```

__Expected result__:
No crash :)

Maybe this bug is related to ArangoJS more than Arangodb server ?
Thanks a lot for your consideration.

Mathias

**Comment by matcho** [comment]
We can confirm that this is fixed in 3.8.3 : crash test program runs fine without crashing the server.

**Comment by jsteemann** [comment]
@matcho: Thanks for the confirmation! I am glad the issue is solved now, and sorry for the trouble!
